File: Ass_1/vatsal_sharma_task2.py
from pyspark.context import SparkContext
from pyspark import SparkConf
from collections import OrderedDict
import time
import sys
import json

conf = SparkConf()
conf.setMaster("local[*]")
sc = SparkContext(conf = conf)
path_review = sys.argv[1]
path_business = sys.argv[2]

output = []

review_file = sc.textFile(path_review)
business_file = sc.textFile(path_business)

rdd_review_json = review_file.map(json.loads).map(lambda x: (x['business_id'],x['stars']))
rdd_business_json = business_file.map(json.loads).map(lambda x: (x['business_id'],x['state']))

rdd_join_review_business = rdd_review_json\
	                       .join(rdd_business_json)\
	                       .map(lambda x : (x[1][1],x[1][0]))\
                           .groupByKey()\
                           .mapValues(lambda x : sum(x) / len(x))\
                           .sortBy(lambda x: (-x[1],x[0]))

list_stars_state = rdd_join_review_business.collect()

text_file = open(sys.argv[3], "a")
text_file.write("state,stars")
text_file.write("\n")
text_file.close()

text_file = open(sys.argv[3], "a")

for value in list_stars_state:
	text_file.write(str(value[0]) + "," + str(value[1]))
	text_file.write("\n")

text_file.close()

start = time.time()
second_method_list = rdd_join_review_business.collect()
for i in range(0,5):
	print(second_method_list[i])
end = time.time() - start

output.append(("m1",end))

start = time.time()
first_methos_list = rdd_join_review_business.take(5)
for value in first_methos_list:
	print(value)
end = time.time() - start

output.append(("m2",end))

explanation = "Method 1 takes more time than method 2 because in method 1 we collect the entire data first and then print the first 5 states where as in method 2 we just take first 5 states from sorted rdd into a list and print it."

output.append(("explanation",explanation))

with open(sys.argv[4], 'w') as outfile:
    json.dump(OrderedDict(output), outfile,indent = 2)

File: Ass_1/vatsal_sharma_task2.scala
import org.apache.spark.SparkContext
import org.apache.spark.sql.{DataFrame, Row, SQLContext, SparkSession}
import java.io._

object vatsal_sharma_task2 {
  def main(args: Array[String])= {

    val spark = SparkSession.builder().master("local[*]").appName("task1").getOrCreate()
    val rdd_review = spark.read.json(args(0)).rdd.map(x => (x.getAs[String]("business_id"),x.getAs[Double]("stars")))
    val rdd_business = spark.read.json(args(1)).rdd.map(x => (x.getAs[String]("business_id"),x.getAs[String]("state")))
    val rdd_join_review_business = rdd_business.join(rdd_review)
    val rdd_state_stars = rdd_join_review_business.map(x => x._2).groupByKey().mapValues(x => x.sum / x.size).sortBy(x => (-x._2,x._1))
    val list_rdd = rdd_state_stars.collect().toList
    val file = args(2)
    val writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(file)))
    writer.write("state,stars" + "\n")
    for (x <- list_rdd) {
      writer.write(x._1.toString() + ',' + x._2.toString() + "\n")  // however you want to format it
    }
    writer.close()

    var start = System.nanoTime
    val list_method_1 = rdd_state_stars.collect()

    for (x <- 0 to 4){
      println(list_method_1(x))
    }
    var end = System.nanoTime
    val time1 = end - start

    start = System.nanoTime
    val list_method_2 = rdd_state_stars.take(5)

    list_method_2.foreach(println)

    end = System.nanoTime

    val time2 = end - start

    val output = scala.collection.immutable.ListMap("m1" -> time1, "m2" -> time2, "explanation" -> "The time here is in nanoseconds. Method 1 takes more time as first we collect the entire data and then print first 5 states whereas in second method we take first 5 states from sorted rdd and print it.")

    var json_object = scala.util.parsing.json.JSONObject(output)

    var file1 = new FileWriter(args(3))
    file1.write(json_object.toString())
    file1.flush()
  }
}

